pennylane/tape/interfaces/autograd.py

```diff
@@ -149,17 +149,73 @@ def get_parameters(self, trainable_only=True, return_arraybox=False):
 
         return params if return_arraybox else autograd.builtins.list(params)
 
-    @autograd.extend.primitive
     def _execute(self, params, device):
-        # unwrap all NumPy scalar arrays to Python literals
-        params = [p.item() if p.shape == tuple() else p for p in params]
-        params = autograd.builtins.tuple(params)
+        """Wrapper to cath the needless autograd forward pass
+
+        This function is not primitive to autograd so when params
+        is a autograd box we know that the following call to
+        the primitve wrapped function is the autograd forward pass.
+        """
+        if isinstance(params, autograd.extend.Box):
+            self.delay_execute_device_as_this_is_a_potentially_needless_autograd_forward_pass = True
+        else:
+            self.delay_execute_device_as_this_is_a_potentially_needless_autograd_forward_pass = False
+        return self.actual_execute(params, device)
+
+    @autograd.extend.primitive
+    def actual_execute(self, params, device):
 
         # unwrap constant parameters
         self._all_params_unwrapped = [
             p.numpy() if isinstance(p, np.tensor) else p for p in self._all_parameter_values
         ]
 
+        if self.delay_execute_device_as_this_is_a_potentially_needless_autograd_forward_pass:
+            self.delay_execute_device_as_this_is_a_potentially_needless_autograd_forward_pass = False
+
+            class DelayedExecutionTensor(np.tensor):
+                """A numpy tensor that delays computation of its content until it is actually needed
+                """
+                def bind(self, *, autograd_interface=None, params=None, device=None):
+                    """Bind this numpy tensor to the current interface instance for delayed execution
+                    """
+                    self.autograd_interface = autograd_interface
+                    self.params = params
+                    self.device = device
+                    self.res = None
+
+                def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):

```

Comment:
The magic happens here.

Reply:
Turns out it was easier to work with `__array_ufunc__` rather than `__array_prepare__`. The current approach may still lead to problems in case the first operation applied to the result is not a ufunc. I am not sure this can happen.